cmd/ts/status.go

```diff
@@ -0,0 +1,355 @@
+}
+
+func NewStatusCmd() *cobra.Command {
+	cmd := &StatusCmd{TsNetFlags: &TsNetFlags{}}
+
+	var statusCmd = &cobra.Command{
+		Use:     "status",
+		Example: "tailscale status [--active] [--web] [--json]",
+		Short:   "Show state of tailscaled and its connections",
+		Args:    cobra.NoArgs,
+		RunE: func(cobraCmd *cobra.Command, args []string) error {
+			ctx := cobraCmd.Context()
+			if len(args) > 0 {
+				return errors.New("unexpected non-flag arguments to 'tailscale status'")
+			}
+
+			// Create network
+			tsNet := tailscale.NewTSNet(&tailscale.TSNetConfig{
+				AccessKey: cmd.AccessKey,
+				Host:      tailscale.RemoveProtocol(cmd.PlatformHost),
+				Hostname:  cmd.NetworkHostname,
+			})
+			// Run tailscale up and wait until we have a connected client
+			done := make(chan bool)
+			go func() {
+				err := tsNet.Start(ctx, done)
+				if err != nil {
+					log.Fatalf("cannot start tsNet server: %v", err)
+				}
+			}()
+			<-done
+
+			// Get tailscale API client
+			localClient, err := tsNet.LocalClient()
+			if err != nil {
+				return fmt.Errorf("cannot get local client: %w", err)
+			}
+			return cmd.runStatus(ctx, localClient)
+		},
+	}
+
+	statusCmd.Flags().BoolVar(&cmd.json, "json", false, "output in JSON format (WARNING: format subject to change)")
+	statusCmd.Flags().BoolVar(&cmd.web, "web", false, "run webserver with HTML showing status")
+	statusCmd.Flags().BoolVar(&cmd.active, "active", false, "filter output to only peers with active sessions (not applicable to web mode)")
+	statusCmd.Flags().BoolVar(&cmd.self, "self", true, "show status of local machine")
+	statusCmd.Flags().BoolVar(&cmd.peers, "peers", true, "show status of peers")
+	statusCmd.Flags().StringVar(&cmd.listen, "listen", "127.0.0.1:8384", "listen address for web mode; use port 0 for automatic")
+	statusCmd.Flags().BoolVar(&cmd.browser, "browser", true, "Open a browser in web mode")
+
+	cmd.ParseFlags(statusCmd)
+
+	return statusCmd
+}
+
+func (cmd *StatusCmd) runStatus(ctx context.Context, localClient *ts.LocalClient) error {
+	getStatus := localClient.Status
+	if !cmd.peers {
+		getStatus = localClient.StatusWithoutPeers
+	}
+	st, err := getStatus(ctx)
+	if err != nil {
+		return fixTailscaledConnectError(err)
+	}
+	if cmd.json {
+		if cmd.active {
+			for peer, ps := range st.Peer {
+				if !ps.Active {
+					delete(st.Peer, peer)
+				}
+			}
+		}
+		j, err := json.MarshalIndent(st, "", "  ")
+		if err != nil {
+			return err
+		}
+		printf("%s", j)
+		return nil
+	}
+	if cmd.web {
+		ln, err := net.Listen("tcp", cmd.listen)
+		if err != nil {
+			return err
+		}
+		statusURL := netmon.HTTPOfListener(ln)
+		printf("Serving Tailscale status at %v ...\n", statusURL)
+		go func() {
+			<-ctx.Done()
+			ln.Close()
+		}()
+		if cmd.browser {
+			go open2.Open(ctx, statusURL, nil)
+		}
+		err = http.Serve(ln, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if r.RequestURI != "/" {
+				http.NotFound(w, r)
```

Comment:
this is definitely way too many nested if statements 😅 can this be somehow refactored into smaller functions?